lib/block-parser.php

```diff
@@ -0,0 +1,205 @@
+<?php
+
+if (!class_exists('Gutenberg_Block_Parser_State', false)):
+
+class Gutenberg_Block_Parser_State {
+	public $block_stack;
+}
+
+endif;
+
+if (!class_exists('Gutenberg_Block_Parser', false)):
+
+class Gutenberg_Block_Parser {
+	const BLOCK_COMMENT_OPEN  = '(^<!--)';
+	const BLOCK_COMMENT_CLOSE = '(^/?-->)';
+	const BLOCK_NAME          = '(^[[:alpha:]](?:[[:alnum:]]|/[[:alnum:]])*)i';
+	const BLOCK_ATTRIBUTES    = '(^{(?:((?!}[ \t\r\n]+/?-->).)*)})';
+	const WS                  = '(^[ \t\r\n])';
+	const WSS                 = '(^[ \t\r\n]+)';
+
+	const MAX_RUNTIME         = 1; // give up after one second
+
+	public function parse($input) {
+		$tic = microtime( true );
+
+		// trampoline for stack-safe recursion of the actual parser
+		while ( $this->input && ( microtime( true ) - $tic ) < self::MAX_RUNTIME ) {
+			return $this->proceed( $input );
+		}
+	}
+
+	public function proceed( $input ) {
+		return succeed( 'test', $input );
+	}
+
+	public static function block_void( $input ) {
+		$result = self::sequence( array(
+			array( 'self::ignore', array( 'self::match', array( '(^<!--[ \t\r\n]+wp:)' ) ) ),
+			array( 'self::match',  array( self::BLOCK_NAME ) ),
+			array( 'self::first_of', array( array(
+				array( 'self::ignore',   array( 'self::match', array( '(^[ \t\r\n]+/-->)' ) ) ),
+				array( 'self::sequence', array( array(
+					array( 'self::ignore', array( 'self::match', array( self::WSS ) ) ),
+					array( 'self::match',  array( self::BLOCK_ATTRIBUTES ) ),
+					array( 'self::ignore', array( 'self::match', array( '(^[ \t\r\n]+/-->)' ) ) )
+				) ) )
+			) ) )
+		), $input );
+
+		if ( empty( $result ) ) {
+			return array();
+		}
+
+		list( list( list( $blockName ), list( list( $raw_attrs ) ) ), $remaining ) = $result;
+		$attrs = $raw_attrs
+			? json_decode( $raw_attrs, true )
+			: array();
+
+		return array( self::block( $blockName, $attrs, '' ), $remaining );
+	}
+
+	//-----------------------------------------
+	// Parser a :: String -> [ ( a, String ) ]
+	//
+	// A parser is a function which takes a string
+	// and returns a list of things and strings
+	//
+	// An empty list is a failed parse
+	//
+	// The polymorphic "a" will eventually be a block
+	//-----------------------------------------
+	public static function succeed( $value, $input ) {
+		return array( array( $value, $input ) );
+	}
+
+	public static function fail( $input ) {
+		return array();
+	}
+
+	public static function ignore( $parser, $parser_args, $input ) {
+		$result = call_user_func_array( $parser, array_merge( $parser_args, array( $input ) ) );
+
+		if ( empty( $result ) ) {
+			return array();
+		}
+
+		list( /* production */, $remaining ) = $result;
+
+		return array( array(), $remaining );
+	}
+
+	public static function literal( $value, $input ) {
+		return strpos( $input, $value ) === 0
+			? array( $value, substr( $input, strlen( $value ) ) )
```

Comment:
Here and below, should we be using `mb_substr` instead?  If not, how can we guarantee that this parser will behave well with UTF-8 input?

If we do switch to `mb_substr`, then performance becomes a concern because it has to scan through the entire string until each start position.